backend/src/main/java/com/melodify/melodify/service/PostService.java

```diff
@@ -0,0 +1,110 @@
+package com.melodify.melodify.service;
+
+import com.melodify.melodify.model.Post;
+import com.melodify.melodify.model.User;
+import com.melodify.melodify.model.Tag;
+import com.melodify.melodify.model.request.PostCreateRequest;
+import com.melodify.melodify.model.request.PostUpdateRequest;
+import com.melodify.melodify.model.response.PostResponse;
+import com.melodify.melodify.repository.PostRepository;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class PostService {
+
+    private final PostRepository postRepository;
+    private final UserService userService;
+    private final TagService tagService;
+
+    public PostService(PostRepository postRepository, UserService userService, TagService tagService) {
+        this.postRepository = postRepository;
+        this.userService = userService;
+        this.tagService = tagService;
+    }
+
+    private List<PostResponse> mapper(List<Post> posts) {
+        return posts.stream().map(post -> {
+            List<String> tags = tagService.getTagsByPost(post).stream().map(Tag::getTag).collect(Collectors.toList());
+            return new PostResponse(post, tags);
+        }).collect(Collectors.toList());
+    }
+
+    public List<PostResponse> getAllPosts() {
+        return mapper(postRepository.findAll());
+    }
+
+    public List<PostResponse> getPostsByAuthorAndTag(String author, String tag) {
+        List<PostResponse> temp = getPostsByAuthor(author);
+        return temp.stream().filter(post -> post.getTags().contains(tag)).collect(Collectors.toList());
+    }
+
+    public List<PostResponse> getPostsByAuthor(String author) {
+        User user = userService.getOneUserByUsername(author);
+        if (user == null) {
+            throw new IllegalArgumentException("User not found");
+        }
+        return mapper(postRepository.findByAuthor(user));
+    }
+
+    public List<PostResponse> getPostsByTag(String tag) {
+        return mapper(tagService.getPostsByTag(tag));
+    }
+
+    public long createOnePost(PostCreateRequest newPostRequest) {
+        User author = userService.getOneUserByUsername(newPostRequest.getAuthor());
```

Comment:
I think it is better to get author info from the authentication token. Otherwise, everyone could create a post for someone else.